packages/siwe-parser/lib/callbacks.ts

```diff
@@ -0,0 +1,482 @@
+    }
+  },
+  uriR: function uriR(result, chars, phraseIndex, data) {
+    switch (result.state) {
+      case id.MATCH:
+        data.uriR = utils.charsToString(
+          chars,
+          phraseIndex,
+          result.phraseLength
+        );
+        break;
+      case id.NOMATCH:
+        data.errors.push(`line ${data.lineno}: invalid resource URI`);
+        break;
+    }
+  },
+  resource: function resource(result, chars, phraseIndex, data) {
+    switch (result.state) {
+      case id.MATCH:
+        if (!data.resources) {
+          data.resources = [];
+        }
+        data.resources.push(data.uriR);
+        delete data.uriR;
+        break;
+      case id.NOMATCH:
+        data.errors.push(`line ${data.lineno}: invalid resource`);
+        break;
+    }
+  },
+  // handle the URI
+  scheme: function scheme(result, chars, phraseIndex, data) {
+    switch (result.state) {
+      case id.MATCH:
+        data.uriElements.scheme = utils.charsToString(
+          chars,
+          phraseIndex,
+          result.phraseLength
+        );
+        break;
+      case id.NOMATCH:
+        data.errors.push(`line ${data.lineno}: invalid URI scheme`);
+        break;
+    }
+  },
+  userinfo: function userinfo(result, chars, phraseIndex, data) {
+    switch (result.state) {
+      case id.MATCH:
+        data.uriElements.userinfo = utils.charsToString(
+          chars,
+          phraseIndex,
+          result.phraseLength - 1
+        );
+        break;
+    }
+  },
+  host: function host(result, chars, phraseIndex, data) {
+    switch (result.state) {
+      case id.ACTIVE:
+        data.iplit = false;
+        break;
+      case id.MATCH:
+        if (data.iplit) {
+          // strip leading "[" and trailing "]" brackets
+          data.uriElements.host = utils.charsToString(
+            chars,
+            phraseIndex + 1,
+            result.phraseLength - 2
+          );
+        } else {
+          data.uriElements.host = utils.charsToString(
+            chars,
+            phraseIndex,
+            result.phraseLength
+          );
+        }
+        break;
+      case id.EMPTY:
+        data.uriElements.host = "";
+        break;
+      case id.NOMATCH:
+        data.errors.push(`line ${data.lineno}: invalid URI host`);
+        break;
+    }
+  },
+  ipLiteral: function ipLiteral(result, chars, phraseIndex, data) {
+    if (result.state === id.MATCH) {
+      data.iplit = true;
+    }
+  },
+  port: function port(result, chars, phraseIndex, data) {
+    switch (result.state) {
+      case id.MATCH:
+        data.uriElements.port = parseIntegerNumber(
+          utils.charsToString(chars, phraseIndex, result.phraseLength)
+        );
+        break;
+      case id.EMPTY:
+        data.uriElements.port = "";
+        break;
+      case id.NOMATCH:
+        data.errors.push(`line ${data.lineno}: invalid URI port`);
+        break;
+    }
+  },
+  pathAbempty: function pathAbempty(result, chars, phraseIndex, data) {
```

Comment:
**Linting errors:** There are only 3 lint problems although they show up in 4 different tests. The first is a trivial warning (see toTrace discussion later) the second is a quick fix of a forgotten `break` statement. However, the file `siwe-parser/lib/siwe-grammar.js` is generated by `apg-js`. So I am in the process of updating `apg-js` (new version 4.4.0) to generate a typescript version of the grammar object file. (This is my first time using typescript, so I'm winging it a bit here.) So pushing any fixes to the fork will have to wait for the new release of `apg-js`. Shouldn't take long.

**packages/siwe-parser/lib/callbacks.ts:** No. The additional callbacks are primarily for validating the URI. They are required for identifying the URI components. Except for obvious errors, no real effort is made to pin down specific errors. For example, a message with `URL: sample.com\n` would fail but you would not get an error message telling you that you had used `URL` instead of `URI`. That's a lot of work and does require a lot of "error productions" with corresponding callbacks.

Reply:
Hi @sbihel. Fixing the `apg-js` generator to generate a typescript-compatible grammar object was not difficult. However, verifying that I didn't break anything in `apg-lite` and especially `apg-js-examples` has turned out to be a little tedious. But I'm almost there. I should have these first two conversations resolved early next week.

Reply:
I just pushed changes to ldthomas/siwe that should resolve all these issues. Let me know if you still have any of these problems.